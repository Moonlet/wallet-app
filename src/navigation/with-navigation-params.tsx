import React from 'react';
import { NavigationScreenProp, NavigationState } from 'react-navigation';
import hoistNonReactStatics from 'hoist-non-react-statics';

export type INavigationProps<P = {}> = P & {
    navigation: NavigationScreenProp<NavigationState, P>;
};

export const withNavigationParams = (options?: { forwardRef?: boolean }) => (Comp: any) => {
    const displayName = `withNavigation(${Comp.displayName || Comp.name || 'Component'})`;
    function Component(props: any) {
        let params;
        if (props.navigation && props.navigation.state && props.navigation.state.params) {
            params = props.navigation.state.params;
        }
        return <Comp {...props} {...params} ref={props.forwardedRef} />;
    }
    Component.displayName = displayName;
    hoistNonReactStatics(Component, Comp);

    function ForwardRef(props, ref) {
        return <Component {...props} forwardedRef={ref} />;
    }

    if (options?.forwardRef) {
        const forwardedRef = React.forwardRef(ForwardRef);

        ForwardRef.displayName = displayName;
        hoistNonReactStatics(forwardedRef, Comp);
        return forwardedRef;
    }
    return Component;
};
