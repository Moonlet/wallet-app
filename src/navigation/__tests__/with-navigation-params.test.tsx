import React from 'react';
import { withNavigationParams } from '../with-navigation-params';
import { shallow } from 'enzyme';

describe('Theme:withTheme', () => {
    describe('withTheme works properly', () => {
        class Comp extends React.Component {
            public static navigationOptions = {
                test: 'OK'
            };

            public static fn() {
                return 'static function';
            }

            public render() {
                return <div />;
            }
        }

        const CompWithNavigationParams: any = withNavigationParams()(Comp);

        test('static properties are transfered', () => {
            expect(CompWithNavigationParams.fn).toBe(Comp.fn);
            expect(CompWithNavigationParams.navigationOptions).toBe(Comp.navigationOptions);
        });

        const testCases = [
            {
                name: 'no navigation prop',
                props: {
                    navigation: undefined
                }
            },
            {
                name: 'navigation without state',
                props: {
                    navigation: {}
                }
            },
            {
                name: 'navigation without params on state',
                props: {
                    navigation: {
                        state: {}
                    }
                }
            },
            {
                name: 'valid navigation',
                props: {
                    navigation: {
                        state: {
                            params: {
                                param1: 'value1',
                                param2: 'value2'
                            }
                        }
                    }
                }
            }
        ];

        for (const testCase of testCases) {
            test(testCase.name, () => {
                const renderedComp = shallow(<CompWithNavigationParams {...testCase.props} />);
                expect(renderedComp.getElement().props).toMatchSnapshot(testCase.name);
            });
        }
    });
});
