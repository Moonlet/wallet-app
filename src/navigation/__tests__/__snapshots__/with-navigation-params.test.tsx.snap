// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme:withTheme withTheme works properly navigation without params on state: navigation without params on state 1`] = `
Object {
  "children": undefined,
  "navigation": Object {
    "state": Object {},
  },
}
`;

exports[`Theme:withTheme withTheme works properly navigation without state: navigation without state 1`] = `
Object {
  "children": undefined,
  "navigation": Object {},
}
`;

exports[`Theme:withTheme withTheme works properly no navigation prop: no navigation prop 1`] = `
Object {
  "children": undefined,
  "navigation": undefined,
}
`;

exports[`Theme:withTheme withTheme works properly valid navigation: valid navigation 1`] = `
Object {
  "children": undefined,
  "navigation": Object {
    "state": Object {
      "params": Object {
        "param1": "value1",
        "param2": "value2",
      },
    },
  },
  "param1": "value1",
  "param2": "value2",
}
`;
